File: tests/r_parity/04_csdid.R
# StatsPAI CS-DiD parity (R side) -- Module 04.
#
# Reads data/04_csdid.csv (the StatsPAI mpdta replica) and runs
# did::att_gt + did::aggte(type="simple"). Tolerance: rel < 1e-3.
#
# Note: did::att_gt's default est_method = "dr" is doubly-robust; we
# use est_method = "reg" to match sp.callaway_santanna's outcome-
# regression doubly-robust variant on this DGP.

.args <- commandArgs(trailingOnly = FALSE)
.file_arg <- grep("^--file=", .args, value = TRUE)
.script_dir <- if (length(.file_arg) > 0) {
  dirname(normalizePath(sub("^--file=", "", .file_arg[1])))
} else {
  getwd()
}
source(file.path(.script_dir, "_common.R"))

suppressPackageStartupMessages({
  library(did)
})

MODULE <- "04_csdid"

df <- read_csv_strict(MODULE)
# did::att_gt internally re-codes never-treated rows to first_treat
# = Inf and silently truncates if the column is integer. Coerce to
# numeric so the never-treated cohort is preserved.
df$first_treat <- as.numeric(df$first_treat)
df$lemp        <- as.numeric(df$lemp)
df$year        <- as.numeric(df$year)

set.seed(PARITY_SEED)
att_gt_fit <- did::att_gt(
  yname    = "lemp",
  tname    = "year",
  idname   = "countyreal",
  gname    = "first_treat",
  data     = df,
  control_group = "nevertreated",
  est_method    = "reg",
  base_period   = "universal",
  bstrap        = FALSE,
  cband         = FALSE
)

agg_simple <- did::aggte(att_gt_fit, type = "simple",
                          bstrap = FALSE, cband = FALSE)

rows <- list(
  parity_row(
    module    = MODULE,
    statistic = "simple_ATT",
    estimate  = agg_simple$overall.att,
    se        = agg_simple$overall.se,
    ci_lo     = agg_simple$overall.att - qnorm(0.975) * agg_simple$overall.se,
    ci_hi     = agg_simple$overall.att + qnorm(0.975) * agg_simple$overall.se,
    n         = nrow(df)
  )
)

# Aggregation vectors. did::aggte returns the cells in $egt with
# $att.egt / $se.egt, and the summary in $overall.att / $overall.se.
for (spec in list(list(type = "dynamic",  label = "event"),
                  list(type = "group",    label = "group"),
                  list(type = "calendar", label = "calendar"))) {
  agg <- did::aggte(att_gt_fit, type = spec$type, bstrap = FALSE, cband = FALSE)
  keys <- agg$egt
  for (i in seq_along(keys)) {
    k <- keys[i]
    key <- if (spec$label == "event" && k >= 0) sprintf("+%d", k) else as.character(k)
    rows[[length(rows) + 1L]] <- parity_row(
      module    = MODULE,
      statistic = sprintf("%s_%s", spec$label, key),
      estimate  = agg$att.egt[i],
      se        = agg$se.egt[i]
    )
  }
  rows[[length(rows) + 1L]] <- parity_row(
    module    = MODULE,
    statistic = sprintf("%s_overall", spec$label),
    estimate  = agg$overall.att,
    se        = agg$overall.se
  )
}

write_results(MODULE, rows,
              extra = list(estimator = "reg",
                           base_period = "universal",
                           control_group = "nevertreated",
                           bstrap = FALSE))

File: tests/r_parity/04_csdid.py
"""StatsPAI CS-DiD parity (Python side) -- Module 04.

Dumps sp.datasets.mpdta() and runs sp.callaway_santanna with the
'reg' (outcome-regression) doubly-robust variant. The companion
04_csdid.R reads the same CSV and runs did::att_gt + did::aggte.

Tolerance: rel < 1e-3 (iterative estimator). The replica is a
deterministic seed=42 simulation calibrated to the published
mpdta neighbourhood.
"""

from __future__ import annotations

import re

import statspai as sp

from _common import ParityRecord, dump_csv, write_results

MODULE = "04_csdid"


def main() -> None:
    df = sp.datasets.mpdta()
    dump_csv(df, MODULE)

    fit = sp.callaway_santanna(
        df,
        y="lemp",
        g="first_treat",
        t="year",
        i="countyreal",
        estimator="reg",
        control_group="nevertreated",
        base_period="universal",
    )

    rows: list[ParityRecord] = [
        ParityRecord(
            module=MODULE,
            side="py",
            statistic="simple_ATT",
            estimate=float(fit.estimate),
            se=float(fit.se),
            ci_lo=float(fit.ci[0]) if fit.ci is not None else None,
            ci_hi=float(fit.ci[1]) if fit.ci is not None else None,
            n=int(len(df)),
        )
    ]

    # Aggregation VECTORS, not just the headline scalar. The event study
    # is the object a parallel-trends claim is read off, and until now
    # this module pinned one number and left the whole path unpinned.
    for agg_type, label in (
        ("dynamic", "event"),
        ("group", "group"),
        ("calendar", "calendar"),
    ):
        agg = sp.aggte(fit, type=agg_type, bstrap=False, cband=False)
        tidy = agg.tidy()
        # dynamic exposes its cells as type="event_study" with terms
        # "event_+1"; group and calendar as type="group_time" with terms
        # "att(g=2004.0)" / "att(t=2005.0)".
        cells = tidy[tidy["type"].isin(("event_study", "group_time"))]
        for term, est, se in zip(cells["term"], cells["estimate"], cells["std_error"]):
            key = re.sub(r"^att\((?:g|t)=", "", str(term)).rstrip(")")
            key = key.replace("event_", "")
            if key.endswith(".0"):
                key = key[:-2]
            rows.append(
                ParityRecord(
                    module=MODULE,
                    side="py",
                    statistic=f"{label}_{key}",
                    estimate=float(est),
                    se=float(se),
                )
            )
        rows.append(
            ParityRecord(
                module=MODULE,
                side="py",
                statistic=f"{label}_overall",
                estimate=float(agg.estimate),
                se=float(agg.se),
            )
        )

    # Stata csdid's `estat group` GAverage holds the cohort shares fixed
    # when it aggregates the per-cohort influence functions, whereas R did
    # and sp.aggte add the share-estimation term (did:::wif). Emit the
    # fixed-share aggregate under its own name so the Stata convention is
    # pinned as a joining row (py <-> Stata) instead of being explained in
    # prose; the default `group_overall` row keeps the did convention.
    agg_fixed = sp.aggte(
        fit, type="group", bstrap=False, cband=False, share_variance=False
    )
    rows.append(
        ParityRecord(
            module=MODULE,
            side="py",
            statistic="group_overall_fixedshare",
            estimate=float(agg_fixed.estimate),
            se=float(agg_fixed.se),
        )
    )

    write_results(
        MODULE,
        "py",
        rows,
        extra={
            "estimator": "reg",
            "control_group": "nevertreated",
            "base_period": "universal",
            "method": fit.method,
            "group_overall_stata_gap": (
                "group_overall is a two-way (StatsPAI/R) match, not a "
                "three-way one: StatsPAI and R did::aggte agree to 1.2e-16; "
                "Stata csdid's estat group GAverage standard error differs by "
                "2.7e-3 (0.27%). Point estimates match on all three sides. "
                "Mechanism: csdid aggregates the per-cohort influence "
                "functions with the cohort shares held fixed, whereas did and "
                "StatsPAI add the share-estimation term (did:::wif). The "
                "group_overall_fixedshare row runs sp.aggte(share_variance="
                "False) and joins csdid's GAverage SE at machine precision "
                "(rel 6e-15), so the convention is pinned rather than "
                "asserted; R did has no fixed-share option, so that row has "
                "no R side by construction."
            ),
            "base_period_note": (
                "base_period is pinned explicitly on BOTH sides. StatsPAI "
                "defaults to 'universal' and R did defaults to 'varying'. "
                "The simple ATT cannot see the difference -- it averages "
                "post-treatment cells only -- so this module passed for "
                "years without the option being matched. The event-study "
                "path can see it, and would have reported a spurious "
                "disagreement on every pre-treatment cell."
            ),
            "se_note": (
                "The simple-ATT point estimate matches R did::aggte and "
                "Stata csdid at rel < 1e-15. The analytic SE now matches the "
                "R/Stata no-bootstrap reference at rel ~ 4e-16, once the "
                "outcome-regression IF carries the control-regression "
                "uncertainty AND sp.aggte carries the cohort-share "
                "weight-estimation term. The registered rel_se tolerance was "
                "1% while that term was missing; it is now 1e-9."
            ),
        },
    )


if __name__ == "__main__":
    main()
